pkg/client/client.go

```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+	response, err := stream.Recv()
+	if err == io.EOF {
+		err = cleanupCacheLockFile(lockFile)
+		if err != nil {
+			return -1, err
+		}
+		os.RemoveAll(tmpObjectDir)
+		return -1, nil
+	}
+	if err != nil {
+		return -1, fmt.Errorf("fs.GetCache receive: %w", err)
+	}
+	version := response.Version
+
+	for _, availableVersion := range availableVersions {
+		if version == availableVersion {
+			return availableVersions[len(availableVersions)-1], nil
+		}
+	}
+
+	tarChan := make(chan *pb.GetCacheResponse, 16)
+	group, ctx := errgroup.WithContext(ctx)
+	ctx, cancel := context.WithCancel(ctx)
+
+	group.Go(func() error {
+		ctx, span := telemetry.Start(ctx, "cache-object-receiver")
+		defer span.End()
+		defer close(tarChan)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case tarChan <- response:
+				response, err = stream.Recv()
+				if err == io.EOF {
+					return nil
+				}
+				if response.Version != version {
+					cancel()
+					return fmt.Errorf("fs.GetCache version mismatch: %d, %d", response.Version, version)
+				}
+				if err != nil {
+					cancel()
+					return fmt.Errorf("fs.GetCache receive: %w", err)
+				}
+			}
+		}
+	})
+
+	for i := 0; i < parallelWorkerCount(); i++ {
+		// create the attribute here when `i` is different
+		attr := key.Worker.Attribute(i)
+
+		group.Go(func() error {
+			ctx, span := telemetry.Start(ctx, "cache-object-writer", trace.WithAttributes(attr))
+			defer span.End()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case response, ok := <-tarChan:
+					if !ok {
+						return nil
+					}
+
+					tarReader := db.NewTarReader(response.Bytes)
+					hashHex := hex.EncodeToString(response.Hash)
+					tempDest := filepath.Join(tmpObjectDir, hashHex)
+					finalDest := filepath.Join(objectDir, hashHex)
+
+					if fileExists(finalDest) {
+						return nil
+					}
+
+					if fileExists(tempDest) {
+						err := os.RemoveAll(tempDest)
+						return fmt.Errorf("temporary cache folder exists for %s and couldn't be removed: %w", tempDest, err)
+					}
+
+					_, err := files.WriteTar(tempDest, CacheObjectsDir(cacheRootDir), tarReader, nil)
+					if err != nil {
+						cancel()
+						return err
+					}
+
+					err = os.Rename(tempDest, finalDest)
+					if err != nil {
+						return fmt.Errorf("couldn't rename temporary folder (%s) to final folder (%s): %w", tempDest, finalDest, err)
+					}
+				}
+			}
+		})
+	}
+
+	err = group.Wait()
+	if err != nil {
+		return -1, err
+	}
+
+	versionFile, err := os.OpenFile(cacheVersionPath(cacheRootDir), os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return -1, fmt.Errorf("fs.GetCache cannot open cache versions file for writing: %w", err)
+	}
+	defer versionFile.Close()
+	if _, err = versionFile.WriteString(fmt.Sprintf("%d\n", version)); err != nil {
+		return -1, fmt.Errorf("fs.GetCache failed to update the versions file: %w", err)
+	}
+
+	err = cleanupCacheLockFile(lockFile)
```

Comment:
So we kind of want to make sure this happens no matter what.

What do you think if we made it so that this function cannot return an error and then add it as a `defer` right after we create the lock.

Trying to avoid cases where we can potentially leave a directory in a bad state that cannot be recovered from.

Reply:
I very much disagree, especially since we currently do not have any recovery mechanisms. 

I'd very much rather have an irrecoverable cache that cannot be used than an inconsistent cache that *may* be recovered from.

In the former case, we're going to be paying more for the transfer of packed files because the cache will be useless (and irrecoverable).

In the latter case, we'd have to deal with multiple, unknown failure modes. We'd have to reason about what all those are before trying to recover from a failure.

Sidenote: I'm starting to think it's harmful to describe the current solution as a "cache", at least from the client's perspective. If there's a cache miss, the application will fail to boot; it will not run in a degraded performance kind of way.